[WIP] 2nd batch of require work: require_relative all the things

milch opened the pull request:
Part 2 of #11244

In this episode we:

- Fixed up `supply`
- Wrote a helper that, for each tool, gives a helper `from_tool` that returns the relative path from the file it is being called from to that tool
- Wrote a custom cop for fastlane that does `require 'tool/module'` -> `require_relative from_tool/'module'`
- Ran that thing on all of fastlane

TODO:

- [ ] Appease the `Style/SpacesAroundOperators` cop because `require_relative from_fastlane/'runner'` looks better than `require_relative from_fastlane / 'runner'`
- [ ] Reenable Require/MissingRequireStatement once I figure out how to make this work with that

milch committed:
Fix supply requires

milch committed:
Add helper for easy relative requires

milch committed:
Auto relative require with rubocop

milch committed:
Refactor to separate file

milch committed:
Disable space around operators for now

milch committed:
Disable missing require statement checker for now

I know they're all there

milch committed:
Run `rubocop -a`

milch committed:
Be more specific with gem cache key (#11504)

* Be more specific with gem cache key

* Also for ubuntu

* Update config.yml

fastlane-bot-helper commented:
<!--
  0 Errors
  2 Warnings: Big PR
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Pull Request is Work in Progress</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

KrauseFx commented on cert/lib/cert/commands_generator.rb:
```diff
@@ -1,7 +1,7 @@
-require 'fastlane/version'
-require 'fastlane_core/configuration/configuration'
-require 'fastlane_core/globals'
+require_relative from_fastlane/'version'
```
`from_fastlane`?

milch replied:
Well  I thought using a function to do this was nicer than `require_relative '../../../fastlane/lib/fastlane/version'`. Ideally I would just name the functions the tool name but that would interfere with the actions in a Fastfile, I believe. So I thought this reads pretty nicely: "require relative from fastlane". We can obviously rename it though. 

Could also replace `require_relative from_fastlane/'version'` (using `Pathname`) with a function `require_relative from_fastlane('version')`.

In practice you can just do `require 'fastlane/version'` anyway and it gets automatically replaced to the require_relative version with the cop: https://github.com/fastlane/fastlane/blob/0a3a16d510c28c44a50e9cea4c9dccf0518bc6d1/rubocop/require_relative.rb#L40

milch replied:
Note that one of the advantages of using `require_relative` for everything is that it minimises cross installation pollution. Since the code is 1/2 converted and 1/2 not converted right now on master it seems to happen more often on broken setups, but once this is done and merged it is impossible for a broken installation (e.g. other versions of fastlane, multiple rubies, old versions of non-monogem gems, ...) to affect fastlane

janpio replied:
FYI: I would never understand that this is a function by myself.

milch replied:
@janpio So you would prefer the `from_fastlane('version')` variant?

milch replied:
The helper with the definitions is here by the way: https://github.com/fastlane/fastlane/pull/11507/files#diff-25e7ffc319c07bcabbd06d6ecc53f9eb

janpio replied:
Yes, definitely. I didn't even notice this was not part of the string, and even if I did I would wonder how it works and be confused. Maybe `require_relative_fastlane` to replace `require_relative` in these cases? That would be clearest to me tbh.

milch replied:
So `require_relative_fastlane 'version'`, `require_relative_spaceship 'portal'`, etc. or `require_relative_fastlane 'fastlane/version'`, `require_relative_fastlane 'spaceship/portal'`? 

Maybe something like `require_internal 'spaceship/portal'` would be more clear as well?

janpio replied:
Yep, that sounds good to me. Then the only thing I could miss is that this is not a native ruby function but something fastlane defines, but actually wouldn't care too much because it is close enough to the native thing in functionality (Is it?).

milch replied:
You'll still be able to do `require 'fastlane/version'`, it would just automatically be replaced by the rubocop when you do `rubocop -a` before submitting, so you don't really need to worry about anything 😄 

Really it just uses the native `require_relative` anyway, but computes the relative path to the other tool for you. Otherwise it can be quite cumbersome to do this manually, and results in long paths like `../../../../fastlane_core/lib/fastlane_core/ui/ui`. This wouldn't be as much of an issue if fastlane didn't have the old directory structure back from when everything used to be separate gems, but right now we have these deep folder trees. 

Maybe once this is complete we can actually do away with the old hierarchy, since changing it will amount to only having to rewrite one function.

janpio replied:
> so you don't really need to worry about anything 😄

Well, I still have to (maybe, at least a bit) understand the code I (or `rubocop -a`) write and commit there...

milch commented:
Did some more preliminary testing, seeing huge speedups with this enabled. 30-40% faster loading... the spinner is basically unnecessary since it disappears instantly

milch committed:
fastlane module refactor

milch committed:
Refactor require_relative helper

Commit:
Add missing require statements for legacy spaceship (PRECHECK) (#11510)

* Add missing require statements for legacy spaceship

* Add missing newline

janpio committed:
Improve "Testing" part of YourFirstPR.md (#11517)

- Add instruction on how to run tests from one specific file
- Add subheadlines to Testing part for more structure
- Change code formatting so alternative testing commands are more prominent

joemasilotti committed:
Bump slack-notifier Gem Dependency (#11465)

* Bump slack-notifier gem dependency

* Account for results coming back as an array

* Continue to build options before initializtion

* Clean up initializers

* Style fixes for gemspec

funnel20 committed:
[frameit] Fix: Empty `keyword` breaks positioning and size of `title` (#11518)

- Ensure that fetch_text() ignores empty strings to prevent that empty text images are created.
- As a result the UI.user_error for missing title is now also shown when an empty title string is defined.
- Improved UI.user_error to also mention title.strings file
- Refactured variable names

jonklo committed:
Remove double whitespace around operators (#11529)

funnel20 committed:
Rename several internal frameit variables (#11532)

The images are a graphical representation of the title and keyword text, so the variables (and 1 function) are renamed to text_image and text_images.

vburojevic committed:
Enable nuking of Enterprise account and warn the user beforehand (#11522)

* Enable nuking of Enterprise account and warn the user beforehand

* Show warning only when nuking enterprise distribution, and exit with error

* Allow nuking of enterprise certificates and provisioning profiles

* Show different warnings when nuking distribution or enterprise

taquitos committed:
Enable Swift FastlaneRunner to be updated automatically 🎉 (#11512)

* Upgrade FastlaneSwiftRunner when needed
- during deployment, generate upgrade manifest
- add readme to autogenerated project

* Only add files to the manifest that the user won't modify

* current logic:
- find all groups in existing project
-- if a group is missing, add it
--- add all files for group into new group, build target, and compile phase
- iterate through existing groups
-- update all files needing updating
- iterate through existing groups
-- if a file from the manifest is new, add it to group, build target, and compile phase
- Save and return true if any action was taken to modify any file (project included)

* Remove dev code that was used during development

* Add header to files people shouldn't modify so we can upgrade em

* Use multiple .reject/select to make more readable

* Don't attempt upgrade at all if running in CI

* Add dry_run to upgrader so we can see if we would make project changes
Ask user if they'd like to upgrade before doing it

* Compare FastlaneRunnerAPIVersion of source and target files
so we don't blindly copy files when not needed
- helps during dry run to see if we need to make changes or not

* Check version numbers of each file and only write out files that need generating
If a generated api needs to be updated, automatically bump the version number
Add failsafe to upgrader, if version numbers are the same but the file content differs, mark for upgrade

* Bump FastlaneAPIVersion in files that were missed last update
Update comments to express users shouldn't modify auto-gen files or distributed code

* Undo test code

janpio committed:
register_devices: replace unsafe external link (#11542)

The link that was hardcoded in the docs doesn't have proper SSL certificates, thus generating a warning in modern browser. The replacement link is `http://`, so no warning.

KrauseFx committed:
We're all one team ❤ (#11491)

AliSoftware committed:
Adding my twitter handle to team.json (#11549)

KrauseFx committed:
Update tty-spinner dependency (#11558)

Fixes https://github.com/fastlane/fastlane/issues/11498
Thanks @piotrmurach

KrauseFx committed:
Update file extension in docs (#11567)

orta committed:
Updates the wording when you include a require statement (#11568)

revolter committed:
Improve team section in README (#11554)

* Fix incorrect jump on anchor click

* Remove conflicting sizes for thumbnails

* Add link to GitHub profile

* Add id for the team table

This assures that adding another table in the README doesn't get
replaced by the generate_team_table Rake task.

KrauseFx committed:
Add link to badge plugin (#11576)

Fixes https://github.com/fastlane/fastlane/issues/11562
Makes it appear on https://docs.fastlane.tools/actions/badge/ also 👍
cc @janpio

revolter committed:
Fix comment reference URL (#11577)

KrauseFx committed:
Don't show version bump info on manual `fastlane init` (#11552)

* Don't show version bump info on manual `fastlane init`

* Fix wrong boolean

KrauseFx committed:
Add link to all contributors in README (#11581)

taquitos committed:
Fix for missing variable in setup_ios.rb (#11587)

Fixes #11584

taquitos committed:
Version bump to 2.76.1 (#11589)

KrauseFx committed:
Update scan test result poster to use new slack API (#11591)

taquitos committed:
Fix for Travis CI not installing WWDRC how we expect it (#11594)

* Fix for Travis CI not installing WWDRC how we expect it
- Travic CI doesn't seem to install the WWDRC cert how we expect
 so attempt to install it, and if we get a failure during install
 check the reason. If it's because the cert exists, then skip.

This is a workaround for #10059

* Using $? instead of status
updated unit test

janpio committed:
Fix spec path separator (#11597)

KrauseFx committed:
Automatic creation + updating of Gemfile on `fastlane init` (#11582)

* Automatic creation + updating of Gemfile on `fastlane init`

* Refactor gemfile code

* Improve output and onboarding

* More improvements

* Remove todo

* Fix code style

emes committed:
Write dSYMs in binary mode to avoid UndefinedConversion error (#11592)

* Write dsym archives in binary mode to avoid Encoding::UndefinedConversionError

* No need for block, just binwrite dSYMs

revolter committed:
Fix broken and old links (#11593)

KrauseFx committed:
Update installation methods across fastlane (#11600)

* Update installation methods across fastlane

* Update tests

KrauseFx reviewed (approved):
Let's do this thing 🚀

KrauseFx commented:
Oh tests 😢

KrauseFx commented:
> the spinner is basically unnecessary since it disappears instantly

OMG 😍

milch commented:
I have no idea what happened here... I'll just close and reopen this one.

janpio commented:
Uh @milch, was this ever reopened or completely abandoned?

milch commented:
Hmm, I remember reopening at some point, but never merging it due to ... issues ... 

In any case, I'm now clear to pick this up again. I'll take a look this weekend!

janpio commented:
As I'm experimenting with performance optimization stuff right now I rebased your branch onto `master`, went through the merge conflicts and created a new PR to see if it actually succeeds in tests: https://github.com/fastlane/fastlane/pull/13458 It does not, and I also left several comment of missing or additional `require`s.
